libqtile/popup/toolkit.py

```diff
@@ -0,0 +1,626 @@
+        ("height", 200, "Height of tooltip"),
+        ("controls", [], "Controls to display"),
+        ("margin", 5, "Margin around edge of tooltip"),
+        ("background", "000000", "Popup background colour"),
+        ("opacity", 1, "Popup window opacity. 'None' inherits bar opacity"),
+        ("close_on_click", True, "Hide the popup when control is clicked")
+    ]
+
+    def __init__(self, qtile, **config):
+        configurable.Configurable.__init__(self, **config)
+        self.add_defaults(_PopupLayout.defaults)
+
+        self.qtile = qtile
+
+        # Define drawable area
+        self._width = self.width - 2 * self.margin
+        self._height = self.height - 2 * self.margin
+
+        # Keep track of which control the mouse is over
+        self.cursor_in = None
+
+        self.configured = False
+
+    def _configure(self):
+        """
+        This method creates an instances of a Popup window which serves as the
+        base for the tooltip.
+
+        We also attach handlers for mouse events so that these can be passed to
+        the relevant controls.
+        """
+        self.popup = Popup(self.qtile,
+                           width=self.width,
+                           height=self.height,
+                           background=self.background,
+                           opacity=self.opacity)
+
+        self.popup.win.process_button_click = self.process_button_click
+        self.popup.win.process_button_release = self.process_button_release
+        self.popup.win.process_pointer_enter = self.process_pointer_enter
+        self.popup.win.process_pointer_leave = self.process_pointer_leave
+        self.popup.win.process_pointer_motion = self.process_pointer_motion
+
+        for c in self.controls:
+            self._place_control(c)
+            c._configure(self.qtile, self)
+
+        self.configured = True
+
+    def _place_control(self, control):
+        """
+        This method should define the offsets and positions for the control.
+
+        Layous therefore need to override this method with the specific rules
+        for that layout.
+        """
+        pass
+
+    def draw(self):
+        """
+        Assuming popup is a fixed size, we can just draw widgets without
+        re-positioning them.
+        """
+        self.popup.clear()
+        for c in self.controls:
+            c.drawer.ctx.save()
+            c.drawer.ctx.translate(c.offsetx, c.offsety)
+            c.paint()
+            c.drawer.ctx.restore()
+        self.popup.draw()
+
+    def show(self, x, y):
+        """Display the tooltip."""
+        if not self.configured:
+            self._configure()
+
+        self.popup.x = x
+        self.popup.y = y
+        self.popup.place()
+        self.popup.unhide()
+        self.draw()
+
+    def hide(self):
+        """Hide the popup."""
+        self.popup.hide()
+
+    def kill(self):
+        self.popup.kill()
+
+    # The below methods are lifted from `bar`
```

Comment:
Perhaps it's worth subclassing Bar? I suppose there is a lot of similarity between the two classes.

Reply:
Yes - may need to look at how feasible that is.

Reply:
Subclassing `Bar` would be overkill. An alternative approach would be to have some sort of mixin that contains these methods and then have both `Bar` and the toolkit inherit from this.